File: solutions/trees/208-implement.js
I'll help you implement a Trie (Prefix Tree) in JavaScript. Here's a complete implementation that meets the requirements:

```javascript
/**
 * Trie (Prefix Tree) Implementation
 * 
 * A Trie is a tree-like data structure used for efficient retrieval of keys in a dataset of strings.
 * Each node in the Trie represents a character, and paths from root to leaf represent complete words.
 */

class TrieNode {
    constructor() {
        // Initialize a map to store child nodes (characters)
        this.children = new Map();
        // Flag to mark if this node represents the end of a word
        this.isEndOfWord = false;
    }
}

class Trie {
    constructor() {
        // Initialize the root node
        this.root = new TrieNode();
    }

    /**
     * Inserts a word into the trie.
     * @param {string} word - The word to insert
     * @return {void}
     */
    insert(word) {
        if (!word) return;

        let current = this.root;
        
        // Traverse through each character in the word
        for (let char of word) {
            // If character doesn't exist in current node's children, create new node
            if (!current.children.has(char)) {
                current.children.set(char, new TrieNode());
            }
            // Move to the next node
            current = current.children.get(char);
        }
        
        // Mark the last node as end of word
        current.isEndOfWord = true;
    }

    /**
     * Returns true if the word is in the trie.
     * @param {string} word - The word to search for
     * @return {boolean}
     */
    search(word) {
        if (!word) return false;

        let current = this.root;
        
        // Traverse through each character in the word
        for (let char of word) {
            if (!current.children.has(char)) {
                return false;
            }
            current = current.children.get(char);
        }
        
        // Return true only if we've reached the end of a word
        return current.isEndOfWord;
    }

    /**
     * Returns true if there is any word in the trie that starts with the given prefix.
     * @param {string} prefix - The prefix to search for
     * @return {boolean}
     */
    startsWith(prefix) {
        if (!prefix) return false;

        let current = this.root;
        
        // Traverse through each character in the prefix
        for (let char of prefix) {
            if (!current.children.has(char)) {
                return false;
            }
            current = current.children.get(char);
        }
        
        // If we've made it through all characters, prefix exists
        return true;
    }
}

// Example usage:
/*
const trie = new Trie();
trie.insert("apple");
console.log(trie.search("apple"));   // returns true
console.log(trie.search("app"));     // returns false
console.log(trie.startsWith("app")); // returns true
trie.insert("app");
console.log(trie.search("app"));     // returns true
*/

// Export the Trie class for use in other modules
module.exports = Trie;
```

This implementation includes:

1. A `TrieNode` class that represents each node in the Trie, containing:
   - A Map to store child nodes
   - A boolean flag to mark if the node represents the end of a word

2. A `Trie` class with three main methods:
   - `insert(word)`: Inserts a new word into the Trie
   - `search(word)`: Searches for a complete word in the Trie
   - `startsWith(prefix)`: Checks if any word in the Trie starts with the given prefix

Key features:
- Efficient prefix-based operations
- Clean and well-commented code
- Proper error handling for edge cases
- ES6+ syntax
- Modular structure with export statement

The implementation is memory-efficient using a Map for storing children, and time-efficient with O(m) complexity for all operations, where m is the length of the word/prefix.

The code handles edge cases such as:
- Empty strings
- Non-existent words/prefixes
- Case sensitivity (maintains input case)

You can use this implementation by importing the Trie class and creating instances as shown in the example usage section.

File: solutions/trees/208-implement.py
I'll help you implement a Trie (Prefix Tree) in Python. Here's a clean and efficient implementation with comments:

```python
class TrieNode:
    """
    A node in the Trie data structure.
    Each node contains:
    - children: dictionary mapping characters to child nodes
    - is_end: boolean indicating if this node represents the end of a word
    """
    def __init__(self):
        self.children = {}  # Maps characters to TrieNode objects
        self.is_end = False  # Marks if this node represents the end of a word


class Trie:
    """
    Implementation of a Trie (Prefix Tree) data structure.
    Supports insertion, search, and prefix search operations.
    """
    
    def __init__(self):
        """
        Initialize the Trie with an empty root node.
        """
        self.root = TrieNode()
    
    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        
        Args:
            word: The string to insert into the trie
        """
        current = self.root
        
        # Traverse the trie, creating new nodes as needed
        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()
            current = current.children[char]
        
        # Mark the end of the word
        current.is_end = True
    
    def search(self, word: str) -> bool:
        """
        Returns True if the word is in the trie, False otherwise.
        
        Args:
            word: The string to search for
            
        Returns:
            bool: True if the word exists in the trie, False otherwise
        """
        current = self.root
        
        # Traverse the trie following the characters in the word
        for char in word:
            if char not in current.children:
                return False
            current = current.children[char]
        
        # Return whether this is a complete word
        return current.is_end
    
    def startsWith(self, prefix: str) -> bool:
        """
        Returns True if there is any word in the trie that starts with the given prefix.
        
        Args:
            prefix: The prefix to search for
            
        Returns:
            bool: True if any word starts with the prefix, False otherwise
        """
        current = self.root
        
        # Traverse the trie following the characters in the prefix
        for char in prefix:
            if char not in current.children:
                return False
            current = current.children[char]
        
        # If we made it through all characters, the prefix exists
        return True


# Example usage
if __name__ == "__main__":
    # Create a new Trie
    trie = Trie()
    
    # Insert some words
    trie.insert("apple")
    trie.insert("app")
    trie.insert("apricot")
    
    # Test the trie
    print(trie.search("apple"))      # True
    print(trie.search("app"))        # True
    print(trie.search("apricot"))    # True
    print(trie.search("apt"))        # False
    print(trie.startsWith("app"))    # True
    print(trie.startsWith("apt"))    # False
```

This implementation includes:

1. A `TrieNode` class that represents each node in the Trie, containing:
   - A dictionary of children nodes
   - A boolean flag indicating if the node represents the end of a word

2. A `Trie` class with three main operations:
   - `insert`: Adds a word to the Trie
   - `search`: Checks if a word exists in the Trie
   - `startsWith`: Checks if any word in the Trie starts with a given prefix

3. Clear comments explaining the implementation and functionality

4. Example usage demonstrating how to use the Trie

The implementation is efficient with the following time complexities:
- Insert: O(m) where m is the length of the word
- Search: O(m) where m is the length of the word
- StartsWith: O(m) where m is the length of the prefix

The code handles edge cases appropriately and follows Python best practices and conventions.